src/main/java/htsjdk/samtools/cram/compression/rans/ransnx16/RANSNx16Decode.java

```diff
@@ -0,0 +1,488 @@
+        final int interleaveSize = (Nway == 4) ? (outSize >> 2) : (outSize >> 5);
+
+        // Number of elements that don't fall into the Nway streams
+        int remSize = outSize - (interleaveSize * Nway);
+        final int out_end = outSize - remSize;
+        for (int i = 0; i < out_end; i += Nway) {
+            for (int r=0; r<Nway; r++){
+
+                // Nway parallel decoding rans states
+                symbols[r] = D.reverseLookup[Utils.RANSGetCumulativeFrequency(rans[r], Constants.TOTAL_FREQ_SHIFT)];
+                outBuffer.put(i+r, symbols[r]);
+                rans[r] = syms[0xFF & symbols[r]].advanceSymbolStep(rans[r], Constants.TOTAL_FREQ_SHIFT);
+                rans[r] = Utils.RANSDecodeRenormalizeNx16(rans[r], inBuffer);
+            }
+        }
+        outBuffer.position(out_end);
+        int reverseIndex = 0;
+
+        // decode the remaining bytes
+        while (remSize>0){
+            byte remainingSymbol = D.reverseLookup[Utils.RANSGetCumulativeFrequency(rans[reverseIndex], Constants.TOTAL_FREQ_SHIFT)];
+            syms[0xFF & remainingSymbol].advanceSymbolNx16(rans[reverseIndex], inBuffer, Constants.TOTAL_FREQ_SHIFT);
+            outBuffer.put(remainingSymbol);
+            remSize --;
+            reverseIndex ++;
+        }
+        outBuffer.rewind();
+        return outBuffer;
+    }
+
+    private ByteBuffer uncompressOrder1WayN(
+            final ByteBuffer inBuffer,
+            final ByteBuffer outBuffer,
+            final RANSNx16Params ransNx16Params) {
+        initializeRANSDecoder();
+
+        // read the first byte and calculate the bit shift
+        final int frequencyTableFirstByte = (inBuffer.get() & 0xFF);
+        final int shift = frequencyTableFirstByte >> 4;
+        final boolean optionalCompressFlag = ((frequencyTableFirstByte & FREQ_TABLE_OPTIONALLY_COMPRESSED_MASK)!=0);
+        ByteBuffer freqTableSource;
+        if (optionalCompressFlag) {
+
+            // spec: The order-1 frequency table itself may still be quite large,
+            // so is optionally compressed using the order-0 rANSNx16 codec with a fixed 4-way interleaving.
+
+            // if optionalCompressFlag is true, the frequency table was compressed using RANS Nx16, N=4 Order 0
+            final int uncompressedLength = Utils.readUint7(inBuffer);
+            final int compressedLength = Utils.readUint7(inBuffer);
+            byte[] compressedFreqTable = new byte[compressedLength];
+
+            // read compressedLength bytes into compressedFreqTable byte array
+            inBuffer.get(compressedFreqTable,0,compressedLength);
+
+            // decode the compressedFreqTable to get the uncompressedFreqTable using RANS Nx16, N=4 Order 0 uncompress
+            freqTableSource = ByteBuffer.allocate(uncompressedLength);
+            ByteBuffer compressedFrequencyTableBuffer = ByteBuffer.wrap(compressedFreqTable);
+            compressedFrequencyTableBuffer.order(ByteOrder.LITTLE_ENDIAN);
```

Comment:
I think the default byte order for new byte buffers is generally BIG_ENDIAN, but I'm fairly certain that the byte order for a buffer that is created by wrapping a byte array is already LITTLE_ENDIAN. We should verify that, and if so, remove this call. More generally, we should try to come up with a  more structured way to set the byte order for these buffers (and for the protocol for rewinding them, etc). We can discuss in the next meeting.